### DISCUSSION TITLE
Understanding Webpack cache contents

### DISCUSSION THREAD

**Discussion by MattIPv4** [webpack/webpack#17185] [Q&A]
👋 We've been running into a pretty consistent issue on Vercel recently where our Next.js cache is over 1.00GB, and, frankly, I have no idea why it is that large. It appears to be mostly due to the Webpack cache files produced during the build.

Looking at Next.js' config for Webpack, it looks like they're doing nothing particularly fancy with the Webpack cache, just enabling it: https://github.com/vercel/next.js/blob/v13.4.2/packages/next/src/build/webpack-config.ts#L2515-L2531

Our Webpack config only makes one further change to the caching, reducing the lifetime to hopefully reduce the cache size by removing unused data: https://github.com/alveusgg/alveusgg/blob/edb7066578727af19d119df14b7719664f06ad58/apps/website/next.config.mjs#L277-L280

The only somewhat non-standard thing we're doing is processing videos in a Webpack loader w/ ffmpeg, but those are only around ~150MB in total, nowhere near close to the 1.00GB limit for the cache.

I have some notes here from my latest testing on this: https://github.com/alveusgg/alveusgg/pull/268#issuecomment-1545049469 -- the tl;dr here is that our entire actual output from the build process for the client/server sits in the ~260MB range, yet the Webpack cache sits at ~580MB when doing a fresh build, and grows if we do a build that uses the previous build's cache.

I have a few questions really that I'd love some help on:

- How can I trace what is in these cache files to understand why they're so large?
- What can I do to optimize the size of the cache, any tips?
- Why does the cache grow on sequential builds (`index.pack.old` files suddenly exist as well as `2.pack` and `3.pack`)? How can I ensure only the latest cache data is being kept on disk?

**Comment by alexander-akait** [comment]
> How can I trace what is in these cache files to understand why they're so large?

It's quite a challenge, but you can try https://webpack.js.org/configuration/cache/#cacheprofile

> What can I do to optimize the size of the cache, any tips?

Compress it https://webpack.js.org/configuration/cache/#cachecompression

> Why does the cache grow on sequential builds (index.pack.old files suddenly exist as well as 2.pack and 3.pack)? How can I ensure only the latest cache data is being kept on disk?

You can set https://webpack.js.org/configuration/cache/#cachemaxage, but I don't think it really helps to solve the problem (and I saw you already tried it)

What is a project size? With images/video/etc? But I agree, this is a big size (but I saw more :smile: ), do you have something experimental for next.js, if yes, can you try to disable it and check again? For example next.js generates a images (with different sizes) on the bundling step, so it will be cached by webpack, also you can just open file using any text editor and found indexes and content of text files

**Reply by MattIPv4** [reply]
> > What can I do to optimize the size of the cache, any tips?
> 
> Compress it https://webpack.js.org/configuration/cache/#cachecompression

The docs linked there say that gzip compression is used by default in production mode, so I assume I'm already benefitting from that?

> > Why does the cache grow on sequential builds (index.pack.old files suddenly exist as well as 2.pack and 3.pack)? How can I ensure only the latest cache data is being kept on disk?
> 
> You can set https://webpack.js.org/configuration/cache/#cachemaxage, but I don't think it really helps to solve the problem (and I saw you already tried it)
> 
> What is a project size? With images/video/etc? But I agree, this is a big size (but I saw more 😄 ), do you have something experimental for next.js, if yes, can you try to disable it and check again? For example next.js generates a images (with different sizes) on the bundling step, so it will be cached by webpack, also you can just open file using any text editor and found indexes and content of text files

Yeah, we tried a max age of just 5 minutes, but even still the output seems to grow. A max age of 0 threw some internal webpack erros, but ideally I'd like to only ever persist what has just been used for the build.

https://github.com/alveusgg/alveusgg/pull/268#issuecomment-1545049469 contains `du -sh` outputs for the build -- looking at the fresh build under the `Testing sequential builds` heading, the entire output from Next.js for our client + server comes to 270MB (`.next/server` + `.next/static`), yet the Webpack cache (`.next/cache/webpack`) is a whopping 583MB.

We are not using any experiment Next.js features beyond scroll restoration, which is just a react-router feature, nothing that'd impact the bundle size.

**Reply by alexander-akait** [reply]
Okay, let's look at files, can you provide commands to build https://github.com/alveusgg/alveusgg and see the problem? Maybe you can create a git branch with mocked `Planetscale` (and other servides) logic?

**Reply by MattIPv4** [reply]
👀 I'll try to create a branch with Planetscale stubbed out and Webpack caching enabled again so you can try it out (we ended up disabling caching and writing our own caching in the video loader as that was really the only thing slowing builds down), later this week probably

**Reply by alexander-akait** [reply]
@MattIPv4 Yeah, feel free to ping me :+1:
